### PR TITLE
[ENH] Speed up evotuning and improve evotuning ergonomics

### LINKED ISSUE #56: Evotuning pairs can be sped up by switching to original NumPy
In informal tests, I swapped out jax NumPy stacking for original NumPy stacking. 

This had a few benefits.

Firstly, `input_output_pairs` was sped up from 4-6 iter/sec to 500 iter/sec. I suspect data transfer between GPU and CPU was the culprit.

FYI the speedups look like this on a little GPU tower with a 8 cores, 16GB, and a GTX1080 (8GB):

image

Secondly, I noticed we might have to set `export XLA_PYTHON_CLIENT_PREALLOCATE=false` before running Jupyter, in order for the environment variable to be set correctly. To be clear, though I did not try setting this environment variable _after_ runtime, as I was trying to chase down the memory issue, so I could be wrong about the flag needing to be set before running Jupyter. @ivanjayapurna this _might_ explain why you still ran into OOM issues even after setting the environment variable on Colab, but I could be wrong here. In any case, though, setting the environment flag before running Jupyter guaranteed that I did not exceed memory usage on the first try.

That said, I still found I had memory accumulation issues. As such, I tried playing around with JIT compilation on CPU and GPU. This seemed to solve the memory issues that we all were experiencing on GPU. In particular, I moved calculation of average loss off from GPU onto CPU, since it's not _that_ computationally intensive (i.e. doesn't involve the backward pass). I also "batched" up a bunch of other calculations, since it appears that computing across the dataset without batching was going to cause memory issues, similar to what happens for image deep learners. 

Some lessons from this exercise:

1. Premature optimization is the root of all evil! I found that preprocessing data using original NumPy, before handing them over to JAX NumPy on GPU, was the right way to go here, as JAX NumPy had a weird `core.py` `bind` thing going on that slowed things down.
2. Memory allocation was a lurking bug waiting to happen, which we could only find by testing on large datasets. Thanks to you, @ivanjayapurna, for putting up the TEM dataset on Colab. I used that productively to find this bug.
3. With all the meetings I have to attend during the day, I find I can't do this kind of deep work during the week. So as a matter of career advice for both of you, _if you're interested in staying deeply technical_ say no to work! :smile_cat: (@ElArkk will know exactly what I'm talking about, he's seen how meetings go at Novartis.)

### PR DESCRIPTION
**Pull request by ericmjl**
## PR Description

Before you read on, please ignore the branch name. I thought originally that I could use numba to speed up things, but it turned out that once again, with some careful profiling, I found we didn't have to.

This PR does a few things:

1. Adds a pre-commit configuration.
2. Adds an installation script that makes easy the installation of jax on GPU.
3. Provided backend specification of device (GPU/CPU).
4. Switched preparation of sequences as input-output pairs exclusively on CPU, for speed.
5. Added ergonomic UI features - progressbars! - that improve user experience.
6. Added docs on recommended batch size and its relationship to GPU RAM consumption.
7. Switched from exact calculation of train/holdout loss to estimated calculation.

In any case, this PR closes #56.

### General
1. [x] I have made the PR off a new branch from my fork 
   (`<your_username>`:`<feature-branch_name>`), *not* 
   `<your_username>`:`master`.
2. [x] I have added my changes to the `CHANGELOG.md` file at the top.
3. [x] I have made any necessary changes to the documentation in the `README`.

### Code checks
1. [x] If there are new features implemented, add suitable tests
   in the `tests` directory.
2. [ ] If any new dependencies are introduced through the new features,
   add the packages, pinned to a version, to `environment.yml`.
3. [x] Run `make test` in a console in the top level directory
   to make sure all the tests pass.
4. [x] Run `make format` in a console in the top level directory
   to make the code comply with the formatting standards.

### CODE REVIEW

**Comment by ericmjl** [comment]
Hmmm, I'm a little confused as to how the code coverage was impacted. I think I need a second opinion on whether stuff could be refactored a bit better. @ElArkk?

**Comment by ElArkk** [comment]
Wow, great work @ericmjl ! Very clever to optionally move average loss computation off of GPU, while retaining it in any case (if available) for the work-intensive weight updates!

As for the coverage, one thing I think could be responsible for the decrease on `evotuning.py` is that we do not supply holdout seqs in the execution test of `evotuning` or `fit`?

File: jax_unirep/evotuning.py [hunk 1]
```diff
@@ -314,37 +352,41 @@ def step(i, state):
...
-        logging.debug("Getting state")
-        state = step(i, state)
+        logger.debug("Getting state")
+        state = step(i, x, y, state)
 
         if i % epoch_len == 0:
             params = get_params(state)
-            loss = avg_loss(xs, ys, params)
+            x, y = len_batching_funcs[l]()
+            loss = avg_loss([x], [y], params, backend=backend)
```
Comment (ElArkk) [inline jax_unirep/evotuning.py:371]: Does this not only use a single batch of sequences instead of all sequences of a given length to calculate the avg loss on the dataset?

Also, is it valid to use just a single length of sequences to calculate avg loss on the dataset when using length batching? In case of random batching, I guess the way we implemented it, it would in any case pass in all sequences because we pad the seqs before we make the `batching_funcs`.

File: jax_unirep/evotuning.py [hunk 2]
```diff
@@ -314,37 +352,41 @@ def step(i, state):
...
         l = choice(seq_lens)
         x, y = len_batching_funcs[l]()
 
         # actual forward & backwrd pass happens here
-        logging.debug("Getting state")
-        state = step(i, state)
+        logger.debug("Getting state")
+        state = step(i, x, y, state)
 
         if i % epoch_len == 0:
```
Comment (ElArkk) [inline jax_unirep/evotuning.py:368]: I think this whole conditional is executed in vain if `loss` is not being used further down in the `steps_per_print` conditional. I guess we should either print the loss here in any case, or move the whole thing down into the `steps_per_print` conditional?

File: jax_unirep/evotuning.py [hunk 3]
```diff
@@ -314,37 +352,41 @@ def step(i, state):
...
+                    f"Estimated average training-set loss: {loss}. "
                     f"Weights dumped."
                 )
 
-                if holdout_seqs is not None:
+                if holdout_seqs:
                     # calculate and print loss for out-domain holdout set
+                    l = choice(holdout_seq_lens)
+                    x, y = holdout_len_batching_funcs[l]()
+                    loss = avg_loss([x], [y], params, backend=backend)
```
Comment (ElArkk) [inline jax_unirep/evotuning.py:386]: Same comment as above about what gets passed to `avg_loss`.

File: jax_unirep/evotuning.py [hunk 4]
```diff
@@ -55,21 +52,46 @@ def evotune_loss(params, inputs, targets):
...
     logging.debug("Calculating average loss.")
     sum_loss = 0
     num_seqs = 0
-    for x, y in zip(xs, ys):
-        sum_loss += evotune_loss(params, inputs=x, targets=y) * len(x)
-        num_seqs += len(x)
+    global evotune_loss  # this is necessary for JIT to reference evotune_loss
+    evotune_loss_jit = jit(evotune_loss, backend=backend)
+
+    def batch_iter(xs: np.ndarray, ys: np.ndarray, batch_size: int = 25):
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply jax_unirep/evotuning.py:80]: Yeah, good catch! We should make it all uniform for simplicity, as there's much less unnecessary complexity this way.

File: jax_unirep/evotuning.py [hunk 5]
```diff
@@ -55,21 +52,46 @@ def evotune_loss(params, inputs, targets):
...
+        i = 0
+        for i in tqdm(
+            range(0, (len(xs)), batch_size),
+            desc=f"Avg loss on dataset length {len(xs)}",
+        ):
+            # while len(xs) > i:
+            yield xs[i : i + batch_size], ys[i : i + batch_size]
+            i += batch_size
+
+    for xmat, ymat in zip(xs, ys):
```
[the remark this answers is not in the document]
Comment (ericmjl) [reply jax_unirep/evotuning.py:90]: At the moment, @ivanjayapurna designed `avg_loss` to take in a list of NumPy arrays, which was a design choice we used when we weren't doing single length. I decided to work with this constraint.

File: jax_unirep/evotuning.py (hunk 1 above)
Comment (ericmjl) [reply jax_unirep/evotuning.py:377]: I changed the docstring to make it clear that we aren't calculating loss on the entire dataset, we are using a random subsample of size `batch_size` to estimate the loss. It's much too expensive to calculate the loss directly; this change improves our ability to iterate.

File: jax_unirep/evotuning.py (hunk 2 above)
Comment (ericmjl) [reply jax_unirep/evotuning.py:374]: Yes, this piece is a bit messy. I think it can be simplified. Let me push something up.

File: jax_unirep/evotuning.py (hunk 3 above)
Comment (ericmjl) [reply jax_unirep/evotuning.py]: Pushed something up. Let me know what you think about the new version.

File: jax_unirep/evotuning.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ElArkk) [reply jax_unirep/evotuning.py:95]: Ok! I realise now that the advantage this way is that `avg_loss` can also be used for mixed length sequences if needed.

**Comment by ericmjl** [comment]
@ivanjayapurna I wanted to get a second eye on the code before we merge. Can you and @ElArkk independently test the `numbjit` branch code in, say, Colab on the GPU runtime? I've been working off my home GPU tower to aid development speed, but I want to make sure I'm not doing something that isn't "generally useful".

**Comment by ElArkk** [comment]
I was just thinking, now that we only use one random batch of sequences to calculate the average loss on the dataset, does it still make sense to have an argument for which backend to use? Since the actual training always uses GPU if available, and needs to be able to handle the same or even larger `batch_size`s as the average loss calculation does. So if GPU memory would be a problem in average loss calc, it will also be one in training anyways? Maybe I'm missing something here @ericmjl ?

**Comment by ericmjl** [comment]
Sometimes, setting the back-end explicitly can help with debugging. For example, in debugging the memory allocation issue, I found it handy to be able to switch freely between CPU and GPU backend. Hence, I think keeping the backend kwarg in there while also setting a sane default (CPU by default)  makes a lot of sense, as it gives both convenience and flexibility.

**Comment by ivanjayapurna** [comment]
Just wanted to add on the testing side - trained on the TEM-1 sequences for 25 epochs on AWS with no memory issues

**Comment by ivanjayapurna** [comment]
Results from 25 epoch training:
107602931_974664602992332_726686476718359154_n

The same results but with "epoch 0" plotted:
107045721_284735082582266_4856007470456448087_n

The plots on the left were prior to random batching, and the plots on the right is results from this PR. It is clear that this made a big difference in the models ability to learn. 2 things:

1.) If I'm understanding it correctly epoch 1 is equivalent to no-evotuning UniRep. Epoch 0 here I believe is the final weights saved without a "step" input to dump_params and is a little confusing. Perhaps we should consider changing this? Simple change, just if no step is inputted its just saved without an index in the name or with "final" appended or something. Perhaps epoch's should be 0 indexed as well, to reflect that epoch 0 is before any training has begun.

2.) There seems to be minimal learning after the 1st epoch still, perhaps the learning rate we used was too high.

**Comment by ElArkk** [comment]
@ivanjayapurna thank you for the thorough testing! What was the learning rate and batch size you used here?

As for `epoch 0`, I'm wondering where it stems from, since the `fit` function does never seem to call on `dump_params` without the `step` argument. But maybe I missed it? 

It could be a good idea to change epoch indexing back to 0, since right now the loss calculations at `epoch 1` actually correspond to 'before any training has happened'. What do you think @ericmjl ?

**Comment by ericmjl** [comment]
> It could be a good idea to change epoch indexing back to 0, since right now the loss calculations at epoch 1 actually correspond to 'before any training has happened'. What do you think @ericmjl ?

Yes, let's do that. I think I messed up the epoch calculation when I did this PR up. @ElArkk do you have a spare cycle to handle it? (If not, no worries, I can get to this later in the week.)

**Comment by ElArkk** [comment]
@ericmjl @ivanjayapurna I did a quick rework of the epoch calculation, let me know if you think it makes sense this way.

**Comment by ericmjl** [comment]
It works for me. Anything else blocking this PR?

**Comment by ElArkk** [comment]
No, I guess this is ready :)

I just have one concern still: Let's say someone wants to evotune on 50-100k sequences. If they use a batch size of ~100, loss per epoch would be calculated on just 0.2% and 0.1% of the whole dataset respectively. Do you think this is enough to infer overall loss @ericmjl ?

**Comment by ericmjl** [comment]
Possibly not for only a few epochs, but in the limit of long-run numbers, it should be not too much of an issue. The key show-stopper that I think we should not compromise on is the interactive feel.

**Comment by ElArkk** [comment]
By interactive feel you mean, not having to wait for too long between epochs for the loss calculation?

In any case, we shouldn't delay merging the sped-up and more stable evotuning any longer! If we see any problems with avg loss calculation, we can always come back to it. @ericmjl what do you think?

**Comment by ericmjl** [comment]
Agreed, hit that button when it’s done!

**Comment by ElArkk** [comment]
Hitting that button after a good nights sleep 😄

**Comment by ericmjl** [comment]
NOICE! (Dude, you have no idea - I was knocked out for 5 hours this afternoon. I’m lacking sleep myself haha.)
